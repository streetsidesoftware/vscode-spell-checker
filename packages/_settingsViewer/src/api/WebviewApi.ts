export interface BaseMessageEvent {
    data: any;
}

export interface Disposable {
    dispose: () => any;
}

export type MessageListener = (e: BaseMessageEvent) => any;

export interface WebviewApi {
    postMessage(msg: any): WebviewApi;
    onmessage: MessageListener | undefined;
    disposable?: Disposable;
}
