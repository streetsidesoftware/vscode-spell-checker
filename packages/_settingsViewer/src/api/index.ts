export * from './message';
export * from './MessageBus';
export * from './WebviewApi';
