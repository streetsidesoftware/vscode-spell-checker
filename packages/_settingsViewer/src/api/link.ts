import { OpenFileMessage } from '.';
import { VsCodeWebviewApi } from './vscode/VsCodeWebviewApi';

export function clickLink(vsCodeApi: VsCodeWebviewApi, event: React.MouseEvent<HTMLAnchorElement>): void {
    event.preventDefault();
    const attrib = event.currentTarget.attributes;
    const href = attrib.getNamedItem('href')?.value;
    if (href) {
        openUri(vsCodeApi, href);
    }
}

export function openUri(vsCodeApi: VsCodeWebviewApi, uri: string): void {
    const msg: OpenFileMessage = {
        command: 'OpenFileMessage',
        value: { uri },
    };
    vsCodeApi.postMessage(msg);
}
