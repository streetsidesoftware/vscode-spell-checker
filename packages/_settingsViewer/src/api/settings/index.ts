export * from './settings';
export * from './workspace';
