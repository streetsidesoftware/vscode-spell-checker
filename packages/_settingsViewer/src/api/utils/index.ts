export function tf(v?: boolean): string {
    return v === true
        ? 'true'
        : v === false
        ? 'false'
        : v
        ? 'true like'
        : v === undefined
        ? 'undefined'
        : v === null
        ? 'null'
        : 'false like';
}

export function uniqueFilter<T>(): (v: T) => boolean {
    const seen = new Set<T>();
    return (v: T) => !!(!seen.has(v) && seen.add(v));
}
