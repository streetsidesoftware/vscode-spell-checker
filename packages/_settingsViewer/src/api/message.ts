import { Settings, ConfigTarget } from './settings';

/**
 * @private
 */
export interface DefinedCommands {
    ConfigurationChangeMessage: ConfigurationChangeMessage;
    EnableLanguageIdMessage: EnableLanguageIdMessage;
    EnableLocaleMessage: EnableLocaleMessage;
    RequestConfigurationMessage: RequestConfigurationMessage;
    SelectFileMessage: SelectFileMessage;
    SelectFolderMessage: SelectFolderMessage;
    SelectTabMessage: SelectTabMessage;
    OpenLinkMessage: OpenLinkMessage;
}

type IsAMessageOf = {
    [k in keyof DefinedCommands]: (msg: IMessage) => msg is DefinedCommands[k];
};

export type Commands = keyof DefinedCommands;

interface IMessage {
    command: Commands;
}

export type CommandMessage = DefinedCommands[Commands];

function isIMessage(data: unknown): data is IMessage {
    if (typeof data !== 'object' || !data || !data.hasOwnProperty('command')) return false;
    return typeof (<IMessage>data).command === 'string';
}

export function isMessage(data: unknown): data is CommandMessage {
    if (!isIMessage(data)) return false;
    const msg = <CommandMessage>data;
    return isMessageOf<CommandMessage>(msg);
}

function isA<T extends IMessage>(cmd: T['command'], fields: [keyof T, (v: any) => boolean][]): (msg: IMessage) => msg is T {
    return function (msg: IMessage): msg is T {
        const t = msg as T;
        return msg.command === cmd && fields.reduce((success: boolean, [key, fn]) => success && fn(t[key]), true);
    };
}

export interface ConfigurationChange {
    activeTab?: string;
    settings: Settings;
}

export interface ConfigurationChangeMessage extends IMessage {
    command: 'ConfigurationChangeMessage';
    value: ConfigurationChange;
}

export interface RequestConfigurationMessage extends IMessage {
    command: 'RequestConfigurationMessage';
}

export interface SelectTabMessage extends IMessage {
    command: 'SelectTabMessage';
    value: string;
}

export interface SelectFolderMessage extends IMessage {
    command: 'SelectFolderMessage';
    value: string;
}

export interface SelectFileMessage extends IMessage {
    command: 'SelectFileMessage';
    value: string;
}

export interface EnableLanguageIdMessage extends IMessage {
    command: 'EnableLanguageIdMessage';
    value: {
        target?: ConfigTarget;
        uri?: string;
        languageId: string;
        enable: boolean;
    };
}

export interface EnableLocaleMessage extends IMessage {
    command: 'EnableLocaleMessage';
    value: {
        target: ConfigTarget;
        uri: string | undefined;
        locale: string;
        enable: boolean;
    };
}

export interface OpenLinkMessage extends IMessage {
    command: 'OpenLinkMessage';
    value: {
        uri: string;
    };
}

export const isConfigurationChangeMessage = isA<ConfigurationChangeMessage>('ConfigurationChangeMessage', [['value', isObject]]);
export const isEnableLanguageIdMessage = isA<EnableLanguageIdMessage>('EnableLanguageIdMessage', [['value', isObject]]);
export const isEnableLocaleMessage = isA<EnableLocaleMessage>('EnableLocaleMessage', [['value', isObject]]);
export const isRequestConfigurationMessage = isA<RequestConfigurationMessage>('RequestConfigurationMessage', []);
export const isSelectFileMessage = isA<SelectFileMessage>('SelectFileMessage', [['value', isString]]);
export const isSelectFolderMessage = isA<SelectFolderMessage>('SelectFolderMessage', [['value', isString]]);
export const isSelectTabMessage = isA<SelectTabMessage>('SelectTabMessage', [['value', isString]]);
export const isOpenLinkMessage = isA<OpenLinkMessage>('OpenLinkMessage', [['value', isObject]]);

function isObject(v: unknown): v is Record<string, unknown> {
    return typeof v === 'object' && v !== null;
}
function isString(v: unknown): v is string {
    return typeof v === 'string';
}

const isMessageOfMap: IsAMessageOf = {
    ConfigurationChangeMessage: isConfigurationChangeMessage,
    EnableLanguageIdMessage: isEnableLanguageIdMessage,
    EnableLocaleMessage: isEnableLocaleMessage,
    RequestConfigurationMessage: isRequestConfigurationMessage,
    SelectFileMessage: isSelectFileMessage,
    SelectFolderMessage: isSelectFolderMessage,
    SelectTabMessage: isSelectTabMessage,
    OpenLinkMessage: isOpenLinkMessage,
};

export function isMessageOf<M extends CommandMessage>(msg: CommandMessage): msg is M {
    const fn = isMessageOfMap[msg.command];
    return fn?.(msg) ?? false;
}
