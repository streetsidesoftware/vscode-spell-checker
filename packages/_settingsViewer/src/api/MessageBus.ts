import { BaseMessageEvent, WebviewApi } from './WebviewApi';
import { Commands, isMessage, CommandMessage } from './message';
import { DefinedCommands, isMessageOf } from '.';

export interface MsgListener<M extends CommandMessage> {
    cmd: M['command'];
    fn: (msg: CommandMessage) => void;
    dispose(): void;
}

type MessageHandler<M extends CommandMessage> = (message: M) => void;

type KnownListeners = {
    [K in keyof DefinedCommands]: MsgListener<DefinedCommands[K]>;
};

type Listener = KnownListeners[keyof KnownListeners];

export interface Messenger {
    listenFor<M extends CommandMessage>(cmd: M['command'], fn: (message: M) => void): Listener;
    postMessage<M extends CommandMessage>(msg: M): void;
}

export type Logger = {
    log: typeof console.log;
    warn: typeof console.warn;
    error: typeof console.error;
    debug: typeof console.debug;
};

export class MessageBus implements Messenger {
    protected listeners = new Map<Commands, Set<Listener>>();

    constructor(readonly vsCodeApi: WebviewApi, public logger: Logger = console) {
        this.vsCodeApi.onmessage = (msg: BaseMessageEvent) => this.respondToMessage(msg);
    }

    listenFor<M extends CommandMessage>(cmd: M['command'], fn: MessageHandler<M>): Listener {
        function handler(msg: CommandMessage) {
            if (isMessageOf<M>(msg)) {
                fn(msg);
            }
        }

        const listener: Listener = {
            fn: handler,
            cmd,
            dispose: () => {
                this.listeners.has(cmd) && this.listeners.get(cmd)!.delete(listener);
            },
        };

        this.listeners.set(cmd, this.listeners.get(cmd) || new Set());
        const cmdListeners = this.listeners.get(cmd)!;
        cmdListeners.add(listener);

        return listener;
    }

    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    postMessage<M extends CommandMessage>(msg: M) {
        this.vsCodeApi.postMessage(msg);
    }

    private respondToMessage(msg: BaseMessageEvent) {
        const message = msg.data;

        if (!isMessage(message)) {
            this.logger.error('Unknown message: %o', msg);
            return;
        }

        const listeners = this.listeners.get(message.command);

        if (!listeners) {
            this.logger.error('Unhandled message: %o', msg);
            return;
        }

        for (const listener of listeners) {
            listener.fn(message);
        }
    }
}
