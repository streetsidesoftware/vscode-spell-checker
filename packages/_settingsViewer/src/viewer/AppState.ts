/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
import { observable, computed, makeObservable, action } from 'mobx';
import { Settings, ConfigTarget, LocaleId, SettingByConfigTarget, WorkspaceFolder, TextDocument, ConfigSource } from '../api/settings/';
import { normalizeCode, lookupCode } from '../iso639-1';
import { compareBy, compareEach } from '../api/utils/Comparable';
import { uniqueFilter } from '../api/utils';
import { Messenger } from '../api';
import { ConfigTargets, configTargets } from '../api/settings/settingsHelper';

type TabTargets = ConfigTarget | 'file' | 'dictionaries' | 'about';

export interface Tab {
    label: string;
    target: TabTargets;
}

const tabs: Tab[] = [
    { label: 'User', target: 'user' },
    { label: 'Workspace', target: 'workspace' },
    { label: 'Folder', target: 'folder' },
    { label: 'File', target: 'file' },
    { label: 'Dictionaries', target: 'dictionaries' },
    { label: 'About', target: 'about' },
];

export interface LanguageInfo {
    code: string;
    name: string;
    dictionaries: string[];
    enabled: boolean;
}

export interface LanguageConfig {
    languages: LanguageInfo[];
    inherited?: ConfigSource;
}

export interface LanguageConfigs extends SettingByConfigTarget<LanguageConfig> {}

export interface State {
    activeTabName: string;
    settings: Settings;
    tabs: Tab[];
    activeTab: Tab;
    languageConfig: LanguageConfigs;
}

export interface FoundInConfig<T> {
    value: Exclude<T, undefined>;
    source: ConfigSource;
}

type InheritedFromSource<T> = {
    value: Exclude<T, undefined>;
    source: ConfigSource;
};

export class AppState implements State {
    @observable _activeTabName = '';
    @observable _settings: Settings = {
        dictionaries: [],
        knownLanguageIds: [],
        configs: {
            user: { inherited: {}, locales: [], languageIdsEnabled: [] },
            workspace: { inherited: { locales: 'user', languageIdsEnabled: 'user' }, locales: [], languageIdsEnabled: [] },
            folder: { inherited: { locales: 'user', languageIdsEnabled: 'user' }, locales: [], languageIdsEnabled: [] },
            file: undefined,
        },
    };
    @observable debugMode: boolean = false;

    constructor(private messageBus: Messenger) {
        makeObservable(this);
    }

    @computed get activeTabName() {
        return this._activeTabName;
    }

    @computed get settings() {
        return this._settings;
    }

    @computed get tabs() {
        const hidden = new Set<TabTargets>(configTargets.filter((target) => !this.settings.configs[target]));
        if (this.workspaceFolders.length <= 1) {
            hidden.add(ConfigTargets.folder);
        }
        if (this.workspaceFolders.length < 1) {
            hidden.add(ConfigTargets.workspace);
        }
        if (!this.settings.activeFileUri) {
            hidden.add('file');
        }
        return tabs.filter((tab) => !hidden.has(tab.target));
    }

    @computed get activeTab() {
        return this.tabs.find((t) => t.label === this.activeTabName) || this.tabs[0];
    }

    @computed get languageConfig(): LanguageConfigs {
        const calcConfig = (target: ConfigTarget): LanguageConfig => {
            const config = this.settings.configs[target];
            const locales = config.locales; // todo: calc inheritance
            const inherited = config.inherited.locales;

            const infos = new Map<string, LanguageInfo>();

            const addLocalesToInfos = (locales: string[], dictionaryName: string | undefined) => {
                locales
                    .map(normalizeCode)
                    .map(lookupCode)
                    .filter(notUndefined)
                    .forEach((lang) => {
                        const { code, lang: language, country } = lang;
                        const name = country ? `${language} - ${country}` : language;
                        const found = this.isLocalEnabledEx(target, code);
                        const enabled = (found && found.value) || false;
                        const info: LanguageInfo = infos.get(name) || {
                            code,
                            name,
                            enabled,
                            dictionaries: [],
                        };
                        if (dictionaryName) {
                            info.dictionaries.push(dictionaryName);
                        }
                        infos.set(name, info);
                    });
            };
            if (locales) {
                addLocalesToInfos(locales, undefined);
            }
            this.settings.dictionaries.forEach((dict) => addLocalesToInfos(dict.locales, dict.name));

            return {
                languages: [...infos.values()].sort(
                    compareEach(
                        compareBy((info) => !info.dictionaries.length),
                        compareBy('name')
                    )
                ),
                inherited,
            };
        };

        return {
            user: calcConfig('user'),
            workspace: calcConfig('workspace'),
            folder: calcConfig('folder'),
        };
    }

    @computed get activeTabIndex(): number {
        const index = this.tabs.findIndex((t) => t.label === this.activeTabName);
        return index > 0 ? index : 0;
    }

    @computed get workspaceFolderNames(): string[] {
        const workspace = this.settings.workspace;
        const folders = (workspace && workspace.workspaceFolders) || [];
        return folders.map((f) => f.name);
    }

    @computed get activeWorkspaceFolder(): string | undefined {
        const folder = this.findMatchingFolder(this.settings.activeFolderUri);
        return folder && folder.name;
    }

    @computed get workspaceFolders(): WorkspaceFolder[] {
        const workspace = this.settings.workspace;
        return (workspace && workspace.workspaceFolders) || [];
    }

    @computed get activeFileUri(): string | undefined {
        return this.settings.activeFileUri;
    }

    @computed get documents(): TextDocument[] {
        const workspace = this.settings.workspace;
        return (workspace && workspace.textDocuments) || [];
    }

    @computed get documentSelection(): { label: string; value: string }[] {
        return this.documents.map((doc) => ({ label: doc.fileName, value: doc.uri }));
    }

    private findMatchingFolder(uri: string | undefined): WorkspaceFolder | undefined {
        return this.workspaceFolders.filter((f) => f.uri === uri)[0];
    }

    private findMatchingFolderByName(name: string | undefined): WorkspaceFolder | undefined {
        return this.workspaceFolders.filter((f) => f.name === name)[0];
    }

    @action actionSetLocale(target: ConfigTarget, locale: LocaleId, enable: boolean) {
        const localesCurrent = this.settings.configs[target].locales;
        const locales = (enable ? [...localesCurrent, locale] : localesCurrent.filter((v) => v !== locale)).filter(uniqueFilter());
        this.settings.configs[target].locales = locales;
        const uri = this.settings.activeFolderUri;
        this.messageBus.postMessage({ command: 'EnableLocaleMessage', value: { target, locale, enable, uri } });
    }

    @action actionSetDebugMode(isEnabled: boolean) {
        this.debugMode = isEnabled;
    }

    private isLocalEnabledEx(field: ConfigTarget, code: LocaleId): InheritedFromSource<boolean> {
        const config = this.settings.configs[field];
        const source = config.inherited.locales || field;
        const locales = config.locales;
        return {
            value: locales.map(normalizeCode).includes(code),
            source,
        };
    }

    @action actionActivateTabIndex(index: number) {
        const tab = this.tabs[index];
        if (tab) {
            this.actionActivateTab(tab.label);
        }
    }

    @action actionActivateTab(tabName: string) {
        this._activeTabName = tabName;
        this.messageBus.postMessage({ command: 'SelectTabMessage', value: this.activeTabName });
    }

    @action actionSelectFolder(folderName: string) {
        const folder = this.findMatchingFolderByName(folderName);
        const folderUri = folder && folder.uri;
        this.settings.activeFolderUri = folderUri;
        if (folderUri !== undefined) {
            this.messageBus.postMessage({ command: 'SelectFolderMessage', value: folderUri });
        }
    }

    @action actionSelectDocument(documentUri: string) {
        this.messageBus.postMessage({ command: 'SelectFileMessage', value: documentUri });
    }

    @action actionEnableLanguageId(languageId: string, enable: boolean, target?: ConfigTarget) {
        const fileConfig = this.settings.configs.file;
        if (!target && fileConfig?.languageId === languageId) {
            fileConfig.languageEnabled = enable;
        }
        const uri = (!target && fileConfig?.uri) || this.settings.activeFolderUri;
        this.messageBus.postMessage({ command: 'EnableLanguageIdMessage', value: { languageId, enable, target, uri } });
    }

    @action updateSettings(settings: Settings): Settings {
        this._settings = settings;
        return this._settings;
    }
}

function notUndefined<T>(a: T): a is Exclude<T, undefined> {
    return a !== undefined;
}
