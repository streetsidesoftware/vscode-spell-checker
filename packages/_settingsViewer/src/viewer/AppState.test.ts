import { toJS } from 'mobx';
import { describe, expect, test } from 'vitest';
import type { Settings } from 'webview-api';

import { MessageBus } from '../api';
import type { WebviewApi } from '../api/WebviewApi';
import { sampleSettings, sampleSettingsSingleFolder } from '../test/samples/sampleSettings';
import { AppState } from './AppState';

// cspell:ignore ripgrep

describe('Validate AppState', () => {
    test('tabs', () => {
        expect(getSampleAppState(sampleSettings).tabs.map((t) => t.target)).toEqual([
            'user',
            'workspace',
            'folder',
            'file',
            'dictionaries',
            'about',
        ]);
        expect(getSampleAppState(sampleSettingsSingleFolder).tabs.map((t) => t.target)).toEqual([
            'user',
            'workspace',
            'file',
            'dictionaries',
            'about',
        ]);
    });

    test('actionActivateTab', () => {
        const appState = getSampleAppState(sampleSettings);
        appState.actionActivateTab('File');
        expect(appState.activeTab).toEqual({ label: 'File', target: 'file' });
        appState.actionActivateTab('');
        expect(appState.activeTab).toEqual({ label: 'User', target: 'user' });
        appState.actionActivateTab('About');
        expect(appState.activeTab).toEqual({ label: 'About', target: 'about' });
    });

    test('actionActivateTabIndex', () => {
        const appState = getSampleAppState(sampleSettings);
        appState.actionActivateTabIndex(3);
        expect(appState.activeTabIndex).toBe(3);
        expect(appState.activeTab).toEqual({ label: 'File', target: 'file' });
        appState.actionActivateTabIndex(0);
        expect(appState.activeTabIndex).toBe(0);
        expect(appState.activeTab).toEqual({ label: 'User', target: 'user' });
        appState.actionActivateTabIndex(5);
        expect(appState.activeTab).toEqual({ label: 'About', target: 'about' });
        appState.actionActivateTabIndex(7);
        expect(appState.activeTabIndex).toBe(5);
        expect(appState.activeTab).toEqual({ label: 'About', target: 'about' });
        appState.actionActivateTabIndex(-1);
        expect(appState.activeTab).toEqual({ label: 'About', target: 'about' });
    });

    test('actionSelectFolder', () => {
        const appState = getSampleAppState(sampleSettings);
        // const uri = 'file:///Users/cspell/projects/clones/ripgrep';
        const folderName = 'ripgrep';
        appState.actionSelectFolder(folderName);
        expect(appState.activeWorkspaceFolder).toBe(folderName);
        appState.actionSelectFolder('unknown');
        expect(appState.activeWorkspaceFolder).toBeUndefined();
    });

    test('languageConfig', () => {
        const appState = getSampleAppState(sampleSettings);
        expect(toJS(appState.languageConfig)).toMatchSnapshot();
    });

    test('workspaceFolderNames', () => {
        const appState = getSampleAppState(sampleSettings);
        expect(toJS(appState.workspaceFolderNames)).toMatchSnapshot();
    });

    function getSampleAppState(settings: Settings): AppState {
        const webviewApi: WebviewApi = {
            postMessage: () => webviewApi,
            onmessage: undefined,
        };
        const msgBus = new MessageBus(webviewApi);
        const appState: AppState = new AppState(msgBus);
        appState.updateSettings(settings);
        return appState;
    }
});
