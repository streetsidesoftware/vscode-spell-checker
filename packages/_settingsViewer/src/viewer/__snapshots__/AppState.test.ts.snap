// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validate AppState languageConfig 1`] = `
Object {
  "folder": Object {
    "inherited": "workspace",
    "languages": Array [
      Object {
        "code": "cs",
        "dictionaries": Array [
          "cs-cz",
        ],
        "enabled": false,
        "name": "Czech",
      },
      Object {
        "code": "en",
        "dictionaries": Array [
          "en_us",
        ],
        "enabled": true,
        "name": "English",
      },
      Object {
        "code": "en-GB",
        "dictionaries": Array [
          "en-gb",
        ],
        "enabled": false,
        "name": "English - United Kingdom",
      },
      Object {
        "code": "en-US",
        "dictionaries": Array [
          "en_us",
        ],
        "enabled": false,
        "name": "English - United States",
      },
      Object {
        "code": "da",
        "dictionaries": Array [],
        "enabled": true,
        "name": "Danish",
      },
    ],
  },
  "user": Object {
    "inherited": undefined,
    "languages": Array [
      Object {
        "code": "cs",
        "dictionaries": Array [
          "cs-cz",
        ],
        "enabled": false,
        "name": "Czech",
      },
      Object {
        "code": "en",
        "dictionaries": Array [
          "en_us",
        ],
        "enabled": true,
        "name": "English",
      },
      Object {
        "code": "en-GB",
        "dictionaries": Array [
          "en-gb",
        ],
        "enabled": false,
        "name": "English - United Kingdom",
      },
      Object {
        "code": "en-US",
        "dictionaries": Array [
          "en_us",
        ],
        "enabled": false,
        "name": "English - United States",
      },
    ],
  },
  "workspace": Object {
    "inherited": undefined,
    "languages": Array [
      Object {
        "code": "cs",
        "dictionaries": Array [
          "cs-cz",
        ],
        "enabled": false,
        "name": "Czech",
      },
      Object {
        "code": "en",
        "dictionaries": Array [
          "en_us",
        ],
        "enabled": true,
        "name": "English",
      },
      Object {
        "code": "en-GB",
        "dictionaries": Array [
          "en-gb",
        ],
        "enabled": false,
        "name": "English - United Kingdom",
      },
      Object {
        "code": "en-US",
        "dictionaries": Array [
          "en_us",
        ],
        "enabled": false,
        "name": "English - United States",
      },
      Object {
        "code": "da",
        "dictionaries": Array [],
        "enabled": true,
        "name": "Danish",
      },
    ],
  },
}
`;

exports[`Validate AppState workspaceFolderNames 1`] = `
Array [
  "ripgrep",
  "dutch",
  "vscode-cspell-dict-extensions",
]
`;
